### PR TITLE
#4487 ContainerPlatform should support custom mode which accepts an e…

### PR DESCRIPTION
**Pull request by chia7712**
…xistent Configurator service

### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4487)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala [hunk 1]
```diff
@@ -130,6 +145,48 @@ object ContainerPlatform {
...
+  /**
+    * custom mode accepts an existent Configurator service to run IT. In this mode, the nodes must be stored in Configurator
+    * already. otherwise, RuntimeException is thrown. The container client is setup according to mode of Configurator.
+    * 1) k8s mode => k8s client
+    * 2) docker mode => docker client
+    * 3) throw RuntimeException
+    *
+    * @return
+    */
+  private[this] def customMode: Option[ContainerPlatform] = sys.env.get(CUSTOM_CONFIGURATOR_KEY).map(_.split(":")).map {
```
Comment (chia7712) [inline ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala:159]: 這一段是關鍵，會找出custom的相關設定，並建立ContainerPlatform

File: ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala [hunk 2]
```diff
@@ -130,6 +145,48 @@ object ContainerPlatform {
...
+          DockerClient(DataCollie(nodes))
+        case _ =>
+          throw new RuntimeException(s"${info.mode} is not supported!!!!")
+      }
+      new ContainerPlatform {
+        override def configuratorHostname: String     = hostname
+        override def configuratorPort: Int            = port
+        override def nodeNames: Set[String]           = nodes.map(_.hostname).toSet
+        override def containerClient: ContainerClient = client
+        override def close(): Unit                    = Releasable.close(client)
```
Comment (chia7712) [inline ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala:181]: k8s and docker模式都會把configurator清除掉，但custom模式則不會

File: ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala (hunk 1 above)
Comment (vitojeng) [reply ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala:157]: Got it.

File: ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala [hunk 3]
```diff
@@ -130,6 +143,53 @@ object ContainerPlatform {
...
+          override def close(): Unit = {
+            // nothing
+          }
+          override lazy val containerClient: ContainerClient = clientCreator()
+        }
+        override def nodeNames: Set[String] = nodes.map(_.hostname).toSet
+        override def toString: String       = "CUSTOM"
+      }
+    case _ =>
+      throw new RuntimeException(
```
Comment (jackyoh) [inline ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala:188]: 這一段看起來像是有設定 CUSTOM_CONFIGURATOR_KEY 參數但是格式不正確

File: ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala (hunk 2 above)
Comment (jackyoh) [reply ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala]: 清除掉的意思是指把 configurator 的 container 清除掉嗎？

**Comment by jackyoh** [comment]
retry

File: ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala (hunk 3 above)
Comment (chia7712) [reply ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala:188]: You are right. We directly throw exception to break test if the value format is wrong.

File: ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala (hunk 2 above)
Comment (chia7712) [reply ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala]: 沒錯，k8s and docker下建立的configurator都會被刪掉

Comment (jackyoh) [reply ohara-it/src/test/scala/oharastream/ohara/it/ContainerPlatform.scala]: OK, 了解。

**Comment by jackyoh** [comment]
retry otherIT

**Comment by chia7712** [comment]
retry otherIT

**Comment by chia7712** [comment]
retry otherIT

**Comment by chia7712** [comment]
retry otherIT

**Comment by chia7712** [comment]
retry connectorIT

**Comment by chia7712** [comment]
retry collieIT

**Comment by vitojeng** [comment]
這個 PR 今天有機會 merge 嗎?

**Comment by chia7712** [comment]
> 這個 PR 今天有機會 merge 嗎?
不急不急, 只是想瞭解一下~ XD

所有測試都過的話...

**Comment by chia7712** [comment]
retry otherIT

**Comment by jackyoh** [comment]
retry collieIT

**Comment by chia7712** [comment]
Let me merge this. I will test all IT later to make sure all work stably ...

**Pull request merged**